core/commands/files/files.go

```diff
@@ -0,0 +1,555 @@
+		ShortDescription: `
+List directories.
+
+Examples:
+
+    $ ipfs files ls /welcome/docs/
+    about
+    contact
+    help
+    quick-start
+    readme
+    security-notes
+
+    $ ipfs files ls /myfiles/a/b/c/d
+    foo
+    bar
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to show listing for"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("l", "use long listing format"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		path := req.Arguments()[0]
+		nd, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		fsn, err := mfs.Lookup(nd.FilesRoot, path)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		switch fsn := fsn.(type) {
+		case *mfs.Directory:
+			listing, err := fsn.List()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			res.SetOutput(&FilesLsOutput{listing})
+			return
+		case *mfs.File:
+			parts := strings.Split(path, "/")
+			name := parts[len(parts)-1]
+			out := &FilesLsOutput{[]mfs.NodeListing{mfs.NodeListing{Name: name, Type: 1}}}
+			res.SetOutput(out)
+			return
+		default:
+			res.SetError(errors.New("unrecognized type"), cmds.ErrNormal)
+		}
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*FilesLsOutput)
+			buf := new(bytes.Buffer)
+			long, _, _ := res.Request().Option("l").Bool()
+
+			for _, o := range out.Entries {
+				if long {
+					fmt.Fprintf(buf, "%s\t%s\t%d\n", o.Name, o.Hash, o.Size)
+				} else {
+					fmt.Fprintf(buf, "%s\n", o.Name)
+				}
+			}
+			return buf, nil
+		},
+	},
+	Type: FilesLsOutput{},
+}
+
+var FilesReadCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Read a file in a given mfs",
+		ShortDescription: `
+Read a specified number of bytes from a file at a given offset. By default, will
+read the entire file similar to unix cat.
+
+Examples:
+
+    $ ipfs files read /test/hello
+    hello
+		`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to file to be read"),
+	},
+	Options: []cmds.Option{
+		cmds.IntOption("o", "offset", "offset to read from"),
+		cmds.IntOption("n", "count", "maximum number of bytes to read"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		path := req.Arguments()[0]
+		fsn, err := mfs.Lookup(n.FilesRoot, path)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		fi, ok := fsn.(*mfs.File)
+		if !ok {
+			res.SetError(fmt.Errorf("%s was not a file", path), cmds.ErrNormal)
+			return
+		}
+
+		offset, _, _ := req.Option("offset").Int()
+
+		_, err = fi.Seek(int64(offset), os.SEEK_SET)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		var r io.Reader = fi
+		count, found, err := req.Option("count").Int()
```

Comment:
i'm in favor of silently returning less. `dd` wont error if you ask it to read more data than it can.